src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -635,10 +651,29 @@ public void Cancel()
             // This object is itself usable as a backing source for ValueTask.  Since there's only ever one awaiter
             // for this object's state transitions at a time, we allow the object to be awaited directly. All functionality
             // associated with the implementation is just delegated to the ManualResetValueTaskSourceCore.
-            private ValueTask GetWaiterTask() => new ValueTask(this, _waitSource.Version);
             ValueTaskSourceStatus IValueTaskSource.GetStatus(short token) => _waitSource.GetStatus(token);
             void IValueTaskSource.OnCompleted(Action<object> continuation, object state, short token, ValueTaskSourceOnCompletedFlags flags) => _waitSource.OnCompleted(continuation, state, token, flags);
             void IValueTaskSource.GetResult(short token) => _waitSource.GetResult(token);
+            private async Task GetWaiterTask(CancellationToken cancellationToken)
+            {
+                var vt = new ValueTask(this, _waitSource.Version).AsTask();
+                using (cancellationToken.Register(s =>
+                {
+                    Http2Stream stream = (Http2Stream)s;
+                    bool signalWaiter;
+                    lock (stream.SyncObject)
+                    {
+                        signalWaiter = stream._hasWaiter;
+                        stream._hasWaiter = false;
+                    }
+                    if (signalWaiter) stream._waitSource.SetException(new OperationCanceledException());
```

Comment:
The synchronization here is making me really nervous.  I don't think this makes things worse, so we shouldn't hold up this PR as a result.  But I think we need to undo the ManualResetValueTaskSource changes I'd previously made and go back to using a TaskCompletionSource for now, and revisit that post-3.0.  Unless I'm missing something, there are subtle race conditions here that could result in multiple threads accessing the MRVTS concurrently, namely one thread calling Reset on it while another calls SetResult/Exception on it, which is not supported and could lead to corruption and hangs.  Unfortunately I don't think we have sufficient testing in place to catch such issues, e.g. I don't believe we have any tests that stress using the same HttpClient instance concurrently against the same server on the same connection and with multiple access patterns.  I will try to make time this weekend after this goes in to submit a PR to at least make it safe, and then we can revisit making it fast.

Reply:
Actually, looking at it more, I think I've convinced myself it is actually safe. The saving grace (which I probably knew when I added the MRVTSC but neglected to comment) is that there can only be one consumer of the waitSource, the code consuming the response and waiting for data to show up. There would be two potential issues:
1. If multiple consumers could race to consume the waitSource
2. If multiple producers could race to set the waitSource
3. If the consumer could race to reset the waitSource concurrently with a producer completing one.

(1) can't happen because there's only ever one consumer. Parallel usage of a response stream is a bad programming error that can result in corruption, but that's the case for most data structures in .NET, and such usage is explicitly not supported (nor easy to accidentally do).

(2) is handled by the lock, which guards  extracting the waitSource and setting the Boolean hasWaiter appropriately.

(3) is the hardest to see, but is addressed again by their only being one consumer. If there could be multiple consumers, then while the producer was calling SetResult outside of the lock, a consumer could come along and Reset it inside of the lock. But since there can only be one consumer, if there is a waiter, by definition it is the sole consumer, which means a consumer couldn't concurrently be calling Reset with SetResult because it's SetResult that would wake up the sole consumer. And if there wasn't a consumer, then hasWaiter would be false, and the producer wouldn't try to SetResult.

I will subsequently add all of this to a comment in the code.

Reply:
Additionally, once the cancellationToken is set, we should never have another waiter (past the one that's cancelled, if there is one). At this point the whole response is cancelled and so no one should be attempting to call GetWaiterTask again. I think this helps, but I'm not entirely sure...  

It might be nice to add an assert about this.

Reply:
thanks @stephentoub for deeper look.